cmd/lncli/cmd_new_macaroon.go

```diff
@@ -0,0 +1,237 @@
+	Name:     "newmacaroon",
+	Category: "Macaroons",
+	Usage: "Bakes a new macaroon with the provided list of " +
+		"permissions and restrictions",
+	ArgsUsage: "[--format= | --save_to=] [--timeout=] [--ip_address=] " +
+		"permissions",
+	Description: `
+	Create a new macaroon that grants the provided permissions and
+	optionally adds restrictions (timeout, IP address) to it.
+
+	The new macaroon can be shown on command line either in JSON
+	(--format=json) or hex (--format=hex) serialized format.
+	Or it can be saved directly to a file using the --save_to argument.
+
+	A permission is a tuple of an entity and an action, separated by a
+	colon. Multiple operations can be added as a comma separated list, for
+	example:
+
+	lncli newmacaroon info:read,invoices:write,foo:bar
+	`,
+	Flags: []cli.Flag{
+		cli.StringFlag{
+			Name: "format",
+			Usage: "the format to serialize and display the " +
+				"macaroon in on the command line, must be " +
+				"either 'json' or 'hex' (default 'hex')",
+		},
+		cli.StringFlag{
+			Name: "save_to",
+			Usage: "save the created macaroon to this file " +
+				"using the default binary format",
+		},
+		cli.Uint64Flag{
+			Name: "timeout",
+			Usage: "the number of seconds the macaroon will be " +
+				"valid before it times out",
+		},
+		cli.StringFlag{
+			Name:  "ip_address",
+			Usage: "the IP address the macaroon will be bound to",
+		},
+		cli.StringFlag{
+			Name: "permissions",
+			Usage: "permissions to include in the macaroon, " +
+				"expressed as a comma separated list of" +
+				"tuples of entity:action",
+		},
+	},
+	Action: actionDecorator(newMacaroon),
+}
+
+func newMacaroon(ctx *cli.Context) error {
+	client, cleanUp := getClient(ctx)
+	defer cleanUp()
+
+	// Show command help if no arguments or flags are provided.
+	if ctx.NArg() == 0 && ctx.NumFlags() == 0 {
+		return cli.ShowCommandHelp(ctx, "newmacaroon")
+	}
+	args := ctx.Args()
+
+	var (
+		format, savePath  string
+		timeout           int64
+		ipAddress         net.IP
+		permissions       []string
+		parsedPermissions []*lnrpc.MacaroonPermission
+		err               error
+	)
+
+	if ctx.String("format") != "" {
+		format = ctx.String("format")
+		if format != "json" && format != "hex" {
+			return fmt.Errorf("format must be either 'json' or " +
+				"'hex'")
+		}
+	}
+
+	if ctx.String("save_to") != "" {
+		savePath = cleanAndExpandPath(ctx.String("save_to"))
+	}
+
+	if ctx.IsSet("timeout") {
+		timeout = ctx.Int64("timeout")
+		if timeout <= 0 {
+			return fmt.Errorf("timeout must be greater than 0")
+		}
+	}
+
+	if ctx.IsSet("ip_address") {
+		ipAddress = net.ParseIP(ctx.String("ip_address"))
+		if ipAddress == nil {
+			return fmt.Errorf("unable to parse ip_address: %s",
+				ctx.String("ip_address"))
+		}
+	}
+
+	// The permissions can either be set through a single positional
+	// argument or through a flag.
+	switch {
+	case ctx.IsSet("permissions"):
+		permissions = strings.Split(ctx.String("permissions"), ",")
```

Comment:
maybe remove the flag, because now the behavior is maybe not as expected when both the flag and the positional args are set.